Evict orphans which are included or precluded by accepted blocks.

gmaxwell opened the pull request:
This eliminates the primary leak that causes the orphan map to
 always grow to its maximum size.

This does not go so far as to attempt to connect orphans made
 connectable by a new block.

Keeping the orphan map less full helps improve the reliability
 of relaying chains of transactions.

pstratem commented on src/main.cpp:
```diff
@@ -2365,6 +2375,14 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+                map<COutPoint, set<uint256> >::iterator itByOutPoint = mapOrphanTransactionsByOutPoint.find(tx.vin[j].prevout);
+                if (itByOutPoint != mapOrphanTransactionsByOutPoint.end()) {
+                    for (set<uint256>::iterator mi = itByOutPoint->second.begin(); mi != itByOutPoint->second.end(); ++mi) {
+                             const uint256& orphanHash = *mi;
```
can you fix the whitespace here

pstratem commented:
ACK 70b5f60c761f56e0e84583e77d20a81db3fe1424

modulo whitespace nit

gmaxwell commented:
So, I'm not super happy with the behavior-- the issue is that it removes included or conflicted orphans, but if those orphans themselves have orphaned children, those won't get removed.   The behavior I'm seeing in testing is that it initially removes many transactions but over time seems to remove fewer or fewer, and I think it's because the orphanmap gets full of double-orphans that were conflicted.

I'm not super keen on the performance implications of recursively eliminating there.

pstratem commented:
@gmaxwell Agreed this behavior is not ideal, but this is an improvement.

sipa commented on src/main.cpp:
```diff
@@ -94,6 +94,7 @@ struct COrphanTx {
 };
 map<uint256, COrphanTx> mapOrphanTransactions GUARDED_BY(cs_main);
 map<uint256, set<uint256> > mapOrphanTransactionsByPrev GUARDED_BY(cs_main);
```
It would be more efficient to store map<uint256, COrphanTx>::iterators instead of uint256's

sipa committed:
Track orphan by prev COutPoint rather than prev hash

sipa commented:
@gmaxwell Here is a commit that switches the mapOrphansByPrev entirely to be by-COutPoint: https://github.com/sipa/bitcoin/tree/reworkorphans

gmaxwell committed:
This eliminates the primary leak that causes the orphan map to
 always grow to its maximum size.

This does not go so far as to attempt to connect orphans made
 connectable by a new block.

Keeping the orphan map less full helps improve the reliability
 of relaying chains of transactions.

gmaxwell committed:
Adds an expiration time for orphan tx.

This prevents higher order orphans and other junk from
 holding positions in the orphan map.  Parents delayed
 twenty minutes are more are unlikely to ever arrive.

The freed space will improve the orphan matching success rate for
 other transactions.

gmaxwell committed:
Increase maximum orphan size to 100,000 bytes.

Although this increases node memory usage in the worst case by perhaps
 30MB, the current behavior causes severe issues with dependent tx relay.

sipa commented on src/main.cpp:
```diff
@@ -689,6 +690,26 @@ void EraseOrphansFor(NodeId peer)
+    if (nNextSweep <= nNow) {
+        // Sweep out expired orphan pool entries:
+        int nErased = 0;
+        int64_t nMinExpTime = nNow + 15 * 60;
```
Can you turn this into a named constant?

sipa commented:
utACK with nit

gmaxwell committed:
Treat orphans as implicit inv for parents, discard when parents rejected.

An orphan whos parents were rejected is never going to connect, so there
 is little utility in keeping it.

Orphans also helpfully tell us what we're missing, so go ahead and treat
 it as INVed.

gmaxwell commented:
@sipa Nit picked.

laanwj commented:
utACK 54326a6
